File: apps/api/src/routes/resume.ts
import { Router } from "express";
import { prisma } from "@placepro/database";
import { authenticate } from "../middleware/auth.js";
import { analyzeResume } from "../services/aiClient.js";

const router = Router();

router.get("/", authenticate, async (req, res, next) => {
  try {
    const resumes = await prisma.resume.findMany({
      where: { userId: req.user!.userId },
      orderBy: { createdAt: "desc" },
    });
    res.json({ success: true, data: resumes });
  } catch (e) {
    next(e);
  }
});

router.post("/analyze", authenticate, async (req, res, next) => {
  try {
    const { fileName, fileUrl } = req.body as { fileName: string; fileUrl: string };
    const userId = req.user!.userId;

    let analysis;
    try {
      analysis = await analyzeResume(fileUrl, userId);
    } catch {
      analysis = {
        ats_score: 72,
        resume_strength: "Good",
        missing_skills: ["Kubernetes", "CI/CD"],
        formatting_issues: ["Use consistent bullet points"],
        keyword_suggestions: ["microservices", "REST API", "agile"],
        improvements: ["Add quantified achievements", "Include GitHub link"],
      };
    }

    const resume = await prisma.resume.create({
      data: {
        userId,
        fileName,
        fileUrl,
        atsScore: analysis.ats_score ?? analysis.atsScore ?? 0,
        analysis,
        suggestions: analysis.improvements ?? analysis.improvement_suggestions,
      },
    });

    await prisma.studentProfile.update({
      where: { userId },
      data: { resumeAtsScore: resume.atsScore ?? 0 },
    });

    res.status(201).json({ success: true, data: resume });
  } catch (e) {
    next(e);
  }
});

export default router;

File: apps/ai-service/routers/resume.py
from fastapi import APIRouter
from pydantic import BaseModel
from services.llm import get_llm_response
from routers import resume_builder

router = APIRouter()
router.include_router(resume_builder.router, tags=["resume-builder"])


class ResumeAnalyzeRequest(BaseModel):
    file_url: str
    user_id: str


@router.post("/analyze")
async def analyze_resume(req: ResumeAnalyzeRequest):
    prompt = f"""Analyze this resume (URL: {req.file_url}) for ATS compatibility.
    Return JSON with: ats_score (0-100), resume_strength, missing_skills (array),
    formatting_issues (array), keyword_suggestions (array), improvements (array)."""

    try:
        result = await get_llm_response(prompt, json_mode=True)
        return result
    except Exception:
        return {
            "ats_score": 75,
            "resume_strength": "Good — clear structure with relevant skills",
            "missing_skills": ["Docker", "Kubernetes", "CI/CD"],
            "formatting_issues": [
                "Inconsistent date formatting",
                "Missing quantified achievements in experience section",
            ],
            "keyword_suggestions": [
                "microservices",
                "REST API",
                "agile methodology",
                "cross-functional",
            ],
            "improvements": [
                "Add metrics to each bullet point (e.g., improved performance by 40%)",
                "Include a dedicated Skills section with job-relevant keywords",
                "Keep resume to 1 page for early career roles",
                "Add GitHub and LinkedIn links prominently",
            ],
        }
